src/db/couchUsers.ts

```diff
@@ -0,0 +1,183 @@
+import { saveToDB } from './utils/couchOps'
+
+export const asUserDevices: Cleaner<UserDevices> = asObject(asBoolean)
+export type IDevicesByCurrencyHoursViewResponse = ReturnType<
+  typeof asUserDevices
+>
+export const asUserCurrencyHours: Cleaner<UserCurrencyHours> = asObject({
+  '1': asBoolean,
+  '24': asBoolean
+})
+export const asUserCurrencyCodes: Cleaner<UserCurrencyCodes> =
+  asMap(asUserCurrencyHours)
+
+export const asUserNotifications: Cleaner<UserNotifications> = asObject({
+  enabled: asOptional(asBoolean),
+  currencyCodes: asUserCurrencyCodes
+})
+
+export const asCouchUser = asCouchDoc<Omit<User, 'userId'>>(
+  asObject({
+    devices: asUserDevices,
+    notifications: asUserNotifications
+  })
+)
+
+type CouchUser = ReturnType<typeof asCouchUser>
+
+// ------------------------------------------------------------------------------
+// Functions associated with User
+// ------------------------------------------------------------------------------
+
+export const registerNotifications = async (
+  userId: string,
+  currencyCodes: string[],
+  connection: ServerScope
+): Promise<void> => {
+  const db = connection.db.use(usersSetup.name)
+  const raw = await db.get(userId).catch(error => {
+    if (asMaybeNotFoundError(error) != null) return
+    throw error
+  })
+  const user = asCouchUser(raw)
+
+  const currencyCodesToUnregister = Object.keys(
+    user.doc.notifications.currencyCodes
+  ).filter(code => !currencyCodes.includes(code))
+  for (const code of currencyCodesToUnregister) {
+    delete user.doc.notifications.currencyCodes[code]
+  }
+
+  for (const code of currencyCodes) {
+    if (code in user.doc.notifications.currencyCodes) continue
+
+    user.doc.notifications.currencyCodes[code] = {
+      '1': true,
+      '24': true
+    }
+  }
+  await saveToDB(db, user, asCouchUser)
+}
+export const devicesByCurrencyHours = async (
+  hours: string,
+  currencyCode: string,
+  connection: ServerScope
+) => {
+  return await connection.db
+    .use(usersSetup.name)
+    .view<IDevicesByCurrencyHoursViewResponse>('filter', 'by-currency', {
+      key: [currencyCode, hours]
+    })
+}
+
+export const attachDevice = async (
+  userId: string,
+  deviceId: string,
+  connection: ServerScope,
+  userDoc?: CouchUser
+): Promise<void> => {
+  const db = connection.db.use(usersSetup.name)
+  const deviceConnection = connection.db.use(devicesSetup.name)
+  const raw = await deviceConnection.get(deviceId).catch(error => {
+    if (asMaybeNotFoundError(error) != null) return
+    throw error
+  })
+  const device = asCouchDevice(raw)
+  if (unpackDevice(device) == null)
+    throw new Error('Device must be registered before attaching to user.')
+  if (userDoc == null) {
+    const raw = await db.get(userId).catch(error => {
+      if (asMaybeNotFoundError(error) != null) return
+      throw error
+    })
+    userDoc = asCouchUser(raw)
+  }
+  userDoc.doc.devices[deviceId] = true
```

Comment:
This is another read / modify / write sequence, so I think this one should also move into the HTTP route.